pkg/ipam/sync.go

```diff
@@ -53,61 +59,110 @@ func (lipam *LiqoIPAM) sync(ctx context.Context, syncFrequency time.Duration) {
+	// Remove networks that are present in the cache but not in the cluster, and were added before the threshold.
+	for currentNetwork := range currentNetworks {
+		if !isNetworkInCluster(clusterNetworks, currentNetwork) {
+			if err := lipam.networkRelease(currentNetwork.String()); err != nil {
+				return fmt.Errorf("failed to free network %q: %w", currentNetwork.String(), err)
+			}
+		}
+	}
+	return nil
+}
+
+func (lipam *LiqoIPAM) syncNetworks(ctx context.Context) error {
 	// List all networks present in the cluster.
-	clusterNetworks, err := listNetworksOnCluster(ctx, lipam.Client)
+	clusterNetworksSlice, err := lipam.listNetworksOnCluster(ctx, lipam.Client)
 	if err != nil {
 		return err
 	}
 
-	// Create the set of networks present in the cluster (for faster lookup later).
-	setClusterNetworks := make(map[string]struct{})
+	clusterNetworksMap := maps.SliceToMap(clusterNetworksSlice)
 
-	// Add networks that are present in the cluster but not in the cache.
-	for _, net := range clusterNetworks {
-		if _, inCache := lipam.cacheNetworks[net.String()]; !inCache {
-			if err := lipam.reserveNetwork(net); err != nil {
-				return err
-			}
+	currentNetworksMap := maps.SliceToMap(lipam.IpamCore.ListNetworks())
+
+	if err := syncNetworkAcquire(lipam, clusterNetworksMap, currentNetworksMap); err != nil {
+		return fmt.Errorf("failed to acquire network: %w", err)
+	}
+
+	if err := syncNetworkFree(lipam, clusterNetworksMap, currentNetworksMap); err != nil {
+		return fmt.Errorf("failed to free network: %w", err)
+	}
+
+	return nil
+}
+
+func syncIPsAcquire(lipam *LiqoIPAM, clusterIPs []ipCidr) {
+	for i := range clusterIPs {
+		_ = lipam.ipAcquireWithAddr(clusterIPs[i].cidr, clusterIPs[i].ip)
+	}
+}
+
+func isIPInCluster(clusterIPs []ipCidr, currentIP ipCidr) bool {
+	for i := range clusterIPs {
+		if currentIP.ip == clusterIPs[i].ip && currentIP.cidr == clusterIPs[i].cidr {
+			return true
 		}
-		setClusterNetworks[net.String()] = struct{}{} // add network to the set
 	}
+	return false
+}
 
-	// Remove networks that are present in the cache but not in the cluster, and were added before the threshold.
-	for key := range lipam.cacheNetworks {
-		if _, inCluster := setClusterNetworks[key]; !inCluster && lipam.cacheNetworks[key].creationTimestamp.Before(expiredThreshold) {
-			lipam.freeNetwork(lipam.cacheNetworks[key].network)
+func syncIPsFreeWithNetwork(lipam *LiqoIPAM, clusterIPs []ipCidr, currentIPs []netip.Addr, currentIPNetwork netip.Prefix) error {
+	for i := range currentIPs {
+		currentIP := currentIPs[i]
+		if !isIPInCluster(clusterIPs, ipCidr{ip: currentIP.String(), cidr: currentIPNetwork.String()}) {
+			if err := lipam.ipRelease(ipCidr{ip: currentIPs[i].String(), cidr: currentIPNetwork.String()}); err != nil {
+				return fmt.Errorf("failed to free IP %q in network %q: %w", currentIPs[i].String(), currentIPNetwork.String(), err)
+			}
 		}
 	}
+	return nil
+}
```

Comment:
Here we do not have optimized the research which is still O(n) instead of O(1). We should cache in a set/map the clusterIPs for a faster lookup